File: core/src/test/java/com/jinloes/prpilot/util/ProcessUtilTest.java
package com.jinloes.prpilot.util;

import static org.assertj.core.api.Assertions.assertThat;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

class ProcessUtilTest {

    @Nested
    class FindBinary {

        @Test
        void firstExistingCandidate_isReturned(@TempDir Path dir) throws IOException {
            Path bin = dir.resolve("mybinary");
            Files.createFile(bin);
            assertThat(ProcessUtil.findBinary("fallback", List.of(bin.toString())))
                    .isEqualTo(bin.toString());
        }

        @Test
        void skipsNonExistentPaths_returnsFirstExisting(@TempDir Path dir) throws IOException {
            Path bin = dir.resolve("real");
            Files.createFile(bin);
            assertThat(
                            ProcessUtil.findBinary(
                                    "fallback",
                                    List.of("/no/such/path", bin.toString(), "/also/missing")))
                    .isEqualTo(bin.toString());
        }

        @Test
        void noCandidateExists_returnsFallbackName() {
            assertThat(
                            ProcessUtil.findBinary(
                                    "mybinary", List.of("/no/such/path", "/another/missing")))
                    .isEqualTo("mybinary");
        }

        @Test
        void emptyCandidateList_returnsFallbackName() {
            assertThat(ProcessUtil.findBinary("mybinary", List.of())).isEqualTo("mybinary");
        }
    }
}

File: core/src/jvmTest/kotlin/com/jinloes/prpilot/util/ProcessUtilTest.kt
package com.jinloes.prpilot.util

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import java.io.File
import java.nio.file.Files

class ProcessUtilJvmTest : FunSpec({

    lateinit var tempDir: File
    beforeTest { tempDir = Files.createTempDirectory("put-test-").toFile() }
    afterTest { tempDir.deleteRecursively() }

    context("findBinary") {

        test("first existing candidate is returned") {
            val bin = File(tempDir, "mybinary").also { it.createNewFile() }
            ProcessUtil.findBinary("fallback", listOf(bin.absolutePath)) shouldBe bin.absolutePath
        }

        test("skips non-existent paths and returns first existing") {
            val bin = File(tempDir, "real").also { it.createNewFile() }
            ProcessUtil.findBinary(
                "fallback",
                listOf("/no/such/path", bin.absolutePath, "/also/missing"),
            ) shouldBe bin.absolutePath
        }

        test("directory is not considered a file — skipped") {
            ProcessUtil.findBinary("fallback", listOf(tempDir.absolutePath)) shouldBe "fallback"
        }
    }
})
